src/TopSurface.h

```diff
@@ -0,0 +1,59 @@
+//Copyright (c) 2017 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#ifndef TOPSURFACE_H
+#define TOPSURFACE_H
+
+#include "utils/polygon.h" //For the polygon areas.
+#include "sliceDataStorage.h" //For the input mesh.
+
+namespace cura
+{
+
+class SliceMeshStorage;
+
+class TopSurface
+{
+public:
+    /*!
+     * Create an empty top surface area.
+     */
+    TopSurface();
+
+    /*!
+     * \brief Generate new top surface for a specific layer.
+     *
+     * The surface will be generated by subtracting the layer above from the
+     * current layer. Anything that is leftover is then part of the top surface
+     * (since there is nothing above it).
+     *
+     * \param mesh The mesh to generate the top surface area for.
+     * \param layer_number The layer to generate the top surface area for.
+     */
+    TopSurface(SliceMeshStorage& mesh, size_t layer_number);
+
+    /*!
+     * \brief The areas of top surface, for each layer.
+     */
+    Polygons areas;
```

Comment:
These top_surface areas are specific to ironing. I think this class should be renamed to `Ironing`.

When we would like to separate out a new feature type called Upper Top Skin, we will not use the areas generated here. I don't see why you have conceptually distinguished between TopSurface area and ironing area in this class.